File: fft/py/fft_perf.py
import math
import time

class Complex:
    def __init__(self, re, im):
        self.re = re
        self.im = im

    def __add__(self, other):
        return Complex(self.re + other.re, self.im + other.im)

    def __sub__(self, other):
        return Complex(self.re - other.re, self.im - other.im)

    def __mul__(self, other):
        return Complex(
            self.re * other.re - self.im * other.im,
            self.re * other.im + self.im * other.re
        )

    def __str__(self):
        return f"{self.re:.2f} + {self.im:.2f}i"

def fft(x):
    n = len(x)
    if n <= 1:
        return

    even = x[0::2]
    odd = x[1::2]

    fft(even)
    fft(odd)

    for k in range(n // 2):
        angle = -2 * math.pi * k / n
        twiddle = Complex(math.cos(angle), math.sin(angle)) * odd[k]
        x[k] = even[k] + twiddle
        x[k + n // 2] = even[k] - twiddle

def main():
    N = 1 << 20
    print(f"Generating {N}-point complex FFT input...")
    print("Technology: Python")

    data = [Complex(i % 256, 0.0) for i in range(N)]

    start = time.perf_counter()
    fft(data)
    end = time.perf_counter()

    print(f"FFT {N} points: {end - start:.6f} seconds")

if __name__ == "__main__":
    main()

File: fft/rust/fft_perf.rs
use std::f64::consts::PI;
use std::time::Instant;

#[derive(Clone, Copy, Debug)]
struct Complex {
    re: f64,
    im: f64,
}

impl Complex {
    fn new(re: f64, im: f64) -> Self {
        Complex { re, im }
    }

    fn add(self, other: Complex) -> Complex {
        Complex::new(self.re + other.re, self.im + other.im)
    }

    fn sub(self, other: Complex) -> Complex {
        Complex::new(self.re - other.re, self.im - other.im)
    }

    fn mul(self, other: Complex) -> Complex {
        Complex::new(
            self.re * other.re - self.im * other.im,
            self.re * other.im + self.im * other.re,
        )
    }
}

fn fft(x: &mut [Complex]) {
    let n = x.len();
    if n <= 1 {
        return;
    }

    let mut even = x.iter().step_by(2).cloned().collect::<Vec<_>>();
    let mut odd = x.iter().skip(1).step_by(2).cloned().collect::<Vec<_>>();

    fft(&mut even);
    fft(&mut odd);

    for k in 0..n / 2 {
        let angle = -2.0 * PI * k as f64 / n as f64;
        let twiddle = Complex::new(angle.cos(), angle.sin()).mul(odd[k]);
        x[k] = even[k].add(twiddle);
        x[k + n / 2] = even[k].sub(twiddle);
    }
}

fn main() {
    const N: usize = 1 << 20;
    println!("Generating {}-point complex FFT input...", N);
    println!("Technology: Rust");

    let mut data = Vec::with_capacity(N);
    for i in 0..N {
        data.push(Complex::new((i % 256) as f64, 0.0));
    }

    let start = Instant::now();
    fft(&mut data);
    let duration = start.elapsed();

    println!("FFT {} points: {:.6} seconds", N, duration.as_secs_f64());
}

